content/blog/2020-10-tokio-0-3.md

```diff
@@ -0,0 +1,129 @@
+
+Below is a simplified version of the new `AsyncRead`:
+
+```rust
+pub trait AsyncRead {
+    fn poll_read(
+        self: Pin<&mut Self>, 
+        cx: &mut Context<'_>, 
+        buf: &mut ReadBuf<'_>
+    ) -> Poll<Result<()>>;
+}
+
+pub struct ReadBuf<'a> {
+    buf: &'a mut [MaybeUninit<u8>],
+    filled: usize,
+    initialized: usize,
+}
+
+impl<'a> ReadBuf<'a> {
+    // functions here, see RFC
+}
+```
+The `&mut [u8]` argument to the Tokio 0.2’s `poll_read` method has some unintended sharp edges: an implementer, not the end consumer, of the `AsyncRead` trait can *read* the data stored in the `&mut [u8]` slice. If a programmer creates a mutable slice of bytes (a `&mut [u8]`) that references uninitialized memory (e.g., the excess capacity in a vector), then *reading* the `&mut [u8]` data would cause undefined behavior. The new design mitigates this issue by providing a `ReadBuf` struct that tracks the uninitialized memory, removing the need to initialize memory.
+
+Additionally, the `poll_read_buf` and `poll_write_buf` methods have been removed from both traits. In practice, these methods were rarely implemented. Vectored operations will be implemented by types that support them.
+
+## Feature flag simplification
+
+We have reduced the amount of feature flags needed to use the core features of Tokio. The `dns`, `tcp`, `udp`,  and `uds` feature flags are collapsed into a single `net` feature flag. We have also combined `rt-core` and `rt-util` into a single `rt` feature flag. This `rt` feature flag now encompasses everything to execute futures except for the multi-threaded runtime which now lives under a `rt-multi-thread` feature flag (previously was called `rt-threaded`).
+
+- `dns`, tcp`, `udp` and `uds` -> `net`
+- `rt-util` and `rt-core` -> `rt`
+- `rt-threaded` -> `rt-multi-threaded`
+
+## Runtime and Builder refactor
+
+The Tokio team has improved the runtime `Builder` to provide better consistency accross feature flag permutations and to be more mis-use resistant. To achieve this, we have added two new constructors to the `Builder` type to support building both variations of the Tokio runtime. Additionally, Tokio will no longer choose the runtime variant based on feature flags when using `Runtime::new` but instead will default to the multi-threaded runtime and is only available under the `rt` feature flag. In addition, we’ve renamed the `core_threads` builder method to the more accurate `worker_threads`. 
+
+We have also refactored the runtime module. Tokio 0.2 exposed two types to interact with the Tokio runtime: `Runtime` and `Handle`. `Runtime` has a `block_on` method that required `&mut self`, whereas `Handle` has a `block_on` method that required `&self`. With Tokio 0.3, we collapsed `Runtime` and `Handle` into a single `Runtime`. Tokio 0.3’s `Runtime`‘s `block_on` now accepts `&self`, which supports concurrent calls to `block_on`. This also means `Runtime` implements `Send` and `Sync` and can therefore be stored in an `Arc`.
```

Comment:
The runtime has always been `Send + Sync`.
```suggestion
We have also refactored the runtime module. Tokio 0.2 exposed two types to interact with the Tokio runtime: `Runtime` and `Handle`. `Runtime` has a `block_on` method that required `&mut self`, whereas `Handle` has a `block_on` method that required `&self`. With Tokio 0.3, we collapsed `Runtime` and `Handle` into a single `Runtime`. The `block_on` method on the new runtime takes `&self`, which means that it can be put in an `Arc` without problems.
```